File: go/problem_42.go
package main

import "fmt"

// subsetSum finds a subset of nums that sums to target using backtracking.
func subsetSum(nums []int, target int) []int {
	var bt func(idx, remaining int, current []int) []int
	bt = func(idx, remaining int, current []int) []int {
		if remaining == 0 { return append([]int{}, current...) }
		if idx == len(nums) || remaining < 0 { return nil }
		// include nums[idx]
		if result := bt(idx+1, remaining-nums[idx], append(current, nums[idx])); result != nil {
			return result
		}
		// skip nums[idx]
		return bt(idx+1, remaining, current)
	}
	return bt(0, target, nil)
}

func problem42() {
	fmt.Println("Problem 42:")
	r1 := subsetSum([]int{12, 1, 61, 5, 9, 2}, 24)
	if r1 == nil { panic("assertion failed") }
	sum := 0
	for _, v := range r1 { sum += v }
	if sum != 24 { panic("assertion failed") }

	r2 := subsetSum([]int{1, 2, 3}, 100)
	if r2 != nil { panic("assertion failed") }

	r3 := subsetSum([]int{5}, 5)
	if len(r3) != 1 || r3[0] != 5 { panic("assertion failed") }
	fmt.Println("All Tests Passed")
}

File: lua/problem_42.lua
local function subsetSum(nums, target)
    local function bt(idx, remaining, current)
        if remaining == 0 then
            local copy = {}
            for _, v in ipairs(current) do copy[#copy+1] = v end
            return copy
        end
        if idx > #nums or remaining < 0 then return nil end
        current[#current+1] = nums[idx]
        local r = bt(idx+1, remaining - nums[idx], current)
        if r then return r end
        current[#current] = nil
        return bt(idx+1, remaining, current)
    end
    return bt(1, target, {})
end

local function problem42()
    print("Problem 42:")
    local r1 = subsetSum({12,1,61,5,9,2}, 24)
    assert(r1 ~= nil, "assertion failed")
    local s = 0; for _, v in ipairs(r1) do s = s + v end
    assert(s == 24, "assertion failed")
    assert(subsetSum({1,2,3}, 100) == nil, "assertion failed")
    local r3 = subsetSum({5}, 5)
    assert(r3 ~= nil and r3[1] == 5, "assertion failed")
    print("All Tests Passed")
end

return { problem42 = problem42 }
